File: 프로그래머스/2/42890. 후보키/후보키.java
import java.util.*;
class Solution {
    public int solution(String[][] relation) {
        HashSet<Integer> columnsSet = new HashSet<>();
        boolean[][] usedColumns = new boolean[(int)Math.pow(2, relation[0].length)][relation[0].length];
        for(int i=0; i<usedColumns.length; i++){
            int num = i;
            for(int j=0; j<usedColumns[0].length; j++){
                usedColumns[i][relation[0].length-1-j] = num % 2 == 1;
                num /= 2;
            }
        }
        for(int i=0; i<usedColumns.length; i++){
            HashSet<String> hs = new HashSet<>();
            for(int j=0; j<relation.length; j++){
                String data = "";
                for(int k=0; k<usedColumns[0].length; k++){
                    if(usedColumns[i][k]){
                        data += relation[j][k];
                    }
                }
                hs.add(data);
            }
            if(hs.size() == relation.length){
                putIn(columnsSet, i);
            }
        }
        int answer = columnsSet.size();
        return answer;
    }
    
    public void putIn(HashSet<Integer> hs, int num){
        if(hs.size() == 0){
            hs.add(num);
            return;
        }
        for(int i : hs){
            if((i & num) == i || (i & num) == num){
                int temp = Math.min(i, num);
                hs.remove(i);
                hs.add(temp);
                return;
            }
        }
        hs.add(num);
    }
}

File: 프로그래머스/2/42890. 후보키/후보키.py
import math
def solution(relation):
    row_len = len(relation)
    col_len = len(relation[0])
    
    def putIn(columnsSet, num):
        if len(columnsSet) == 0:
            columnsSet.add(num)
        else:
            for i in columnsSet:
                if (i & num) == i or (i & num) == num:
                    num = min(i, num)
                    columnsSet.remove(i)
                    columnsSet.add(num)
            columnsSet.add(num)
                
    columnsSet = []
    
    for i in range(1, 1 << col_len):
        dataSet = set()
        for r in range(row_len):
            data = []
            for c in range(col_len):
                if i & (1 << c):
                    data.append(relation[r][c])
            dataSet.add(tuple(data))
        if len(dataSet) == len(relation):
            columnsSet.append(i)

    columnsSet.sort(key=lambda x: bin(x).count('1'))
    finalKeys = set()
    for i in range(len(columnsSet)):
        is_minimal = True
        for j in range(i):
            if (columnsSet[j] & columnsSet[i]) == columnsSet[j]:
                is_minimal = False
                break
        if is_minimal:
            finalKeys.add(columnsSet[i])
            
    answer = len(finalKeys)
    return answer